craft_parts/overlays/chroot.py

```diff
@@ -0,0 +1,148 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Execute a callable in a chroot environment."""
+
+import logging
+import multiprocessing
+import os
+import sys
+from collections import namedtuple
+from multiprocessing.connection import Connection
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Tuple
+
+from craft_parts.utils import os_utils
+
+logger = logging.getLogger(__name__)
+
+
+def chroot(path: Path, target: Callable, *args, **kwargs) -> Any:
+    """Execute a callable in a chroot environment.
+
+    :param path: The new filesystem root.
+    :param target: The callable to run in the chroot environment.
+    :param args: Arguments for target.
+    :param kwargs: Keyword arguments for target.
+
+    :returns: The target function return value.
+    """
+    logger.debug("[pid=%d] parent process", os.getpid())
+    parent_conn, child_conn = multiprocessing.Pipe()
+    child = multiprocessing.Process(
+        target=_runner, args=(Path(path), child_conn, target, args, kwargs)
+    )
+    _setup_chroot(path)
+    try:
+        child.start()
+        res, exc = parent_conn.recv()
+        child.join()
+    finally:
+        _cleanup_chroot(path)
+
+    if isinstance(exc, Exception):
+        raise exc
+
+    return res
+
+
+def _runner(
+    path: Path,
+    conn: Connection,
+    target: Callable,
+    args: Tuple,
+    kwargs: Dict,
+) -> None:
+    """Chroot to the execution directory and call the target function."""
+    logger.debug("[pid=%d] child process: target=%r", os.getpid(), target)
+    try:
+        logger.debug("[pid=%d] chroot to %r", os.getpid(), path)
+        os.chdir(path)
+        os.chroot(path)
+        res = target(*args, **kwargs)
+    except Exception as exc:  # pylint: disable=broad-except
+        conn.send((None, exc))
+        return
+
+    conn.send((res, None))
+    sys.exit()
+
+
+def _setup_chroot(path: Path) -> None:
+    """Prepare the chroot environment before executing the target function."""
+    logger.debug("setup chroot: %r", path)
+    if sys.platform == "linux":
+        _setup_chroot_linux(path)
```

Comment:
```suggestion
    if sys.platform != "linux":
        raise ....

    _setup_chroot_linux(path)
```

Reply:
the easy way to guard from this at a higher level is to make the overlay keywords error if not on linux, but this will pass silently (nothing happens effect)

Reply:
Yes, we need a higher level guard.

Reply:
Will address this issue in a follow-up PR.